contributors/devel/sig-architecture/api-conventions.md

```diff
@@ -900,8 +900,34 @@ clearly described and the permissions issues should be resolved.
 This could be done with a double opt-in (an opt-in from both the referrer and the refer-ee) or with secondary permissions
 checks performed in admission. 
 
-TODO: Plugins, extensions, nested kinds, headers
+TODO: Plugins, extensions, headers
 
+### Handling references to multiple kinds
+
+References which can refer to multiple kinds should use a single field, and select the target kind via `apiVersion` and `kind` fields.
+
+For example, if one can retrieving a referenced value from a `ConfigMap` or a `Secret` kind, the schema for the reference should be of the form:
+
+```yaml
+# preferred pattern
+valueFrom:
+  kind: Secret  # alternatively ConfigMap
+  name: resource-name
+  optional: true
+```
+
+Rather than:
+
+```yaml
+# discouraged pattern
+valueFrom:
+  configMapRef:
```

Comment:
This pattern makes it obvious to the user which resources are supported. If you don't mean to generically support arbitrary things, is this really worse?

Reply:
I think for consistencies sake, recommending one (kind) or the other (nested refs) would be good.

IMO, interning kind information into the field name is problematic because:

* tools like kustomize or kpt don't have great mechanisms for dealing with those.
  * kustomize IIUC would have to do one addition and one remove operation with nearly identical fields to change the referred kind
  * kpt has no facility for this at all, requiring multiple blueprints, one for each kind variation
* policy tools like opa would need one rule per sub-field to apply a generic rule to a reference field
* instead of knowing the kind name and pasting that in, you have to guess (to some degree) how the schema chooses to normalize the kind name. (poor example, but something like `configMapRef` vs `configmapRef`).

Reply:
> do one addition and one remove operation

This is a general problem with union/one-of statements; we can now solve this everywhere with a followup to SSA.

> policy tools like opa would need one rule per sub-field to apply a generic rule

Can you give an example such rule?

> instead of knowing the kind name and pasting that in, you have to guess (to some degree) how the schema chooses to normalize the kind name

eh, if a human can figure out what it means, that is good enough, I would not expect code that expects to mechanically turn a field name into a kind/resource name-- I'd expect a lookup table.

Reply:
> I would not expect code that expects to mechanically turn a field name into a kind/resource name-- I'd expect a lookup table.

Yes, I don't want machinery to parse the field name, +1.

Reply:
There's another reason I forgot too: it makes the schema very similar to that of a field reference as well, further normalizing the schema. 

> This is a general problem with union/one-of statements; we can now solve this everywhere with a followup to SSA.

> eh, if a human can figure out what it means, that is good enough, I would not expect code that expects to mechanically turn a field name into a kind/resource name-- I'd expect a lookup table.

Although there may be solutions or they may be reasoned against, they are legitimate issues that make the API harder to use.

Is there any pro to using this style of syntax? 

> Can you give an example such rule?

Sure. [here is a link to the section in the google doc explaining it](https://docs.google.com/document/d/1N2MbLsO0xLCa4PZkEM8g3Skt7XKv8C49_6yW1gBiMic/edit#heading=h.2icpvoiy8qa). Inlining for reference:

--- 

The multivariate keys require policy enforcement to iterate through all of the possible sub-values (e.g. [targetGRPCProxyRef in Config Connector](https://cloud.google.com/config-connector/docs/reference/resource-docs/compute/computeforwardingrule)), making simple policy such as “no references can go cross-namespace” require iteration through all the values. 

As an example, with a single kind, the policy becomes:

```
not spec.target.namespace
```

While with multiple field names it becomes:

```
not spec.target.targetHTTPProxyRef.namespace
not spec.target.targetHTTPSProxyRef.namespace
not spec.target.targetSSLProxyRef.namespace
not spec.target.targetVPNGatewayRef.namespace
not spec.target.targetSSLProxyRef.namespace
```

---

Reply:
It looks like something like `not spec.target[_].namespace` would work in Rego? (https://www.openpolicyagent.org/docs/latest/policy-language/)

But I still think it's a good example. However I feel like there's a counter example that has been made harder, e.g. if you want to forbid a particular kind of reference, or a reference to a sub field within some type(s) of object, now you have to do a switch statement or something similar.

Reply:
> e.g. if you want to forbid a particular kind of reference, or a reference to a sub field within some type(s) of object, now you have to do a switch statement or something similar.

Agreed that some cases would be harder, but I think it's not *too* bad. Forbidding a specific resource would be:

```
not spec.target.resource == "fooresource"
```

unify that with some other condition would be:

```
default allow = false

allow { 
   not spec.target.resource == "fooresource"
}

allow {
    spec.target.resource == "fooresource"
    not spec.target.namespace
}
```

I may try to look a bit more into rego policies that would be negatively impacted by this syntax, but I think in general value-based parameterization is a little cleaner (simpler value comparison vs not / existing statements)